src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -7559,7 +7559,12 @@ and TcForEachExpr cenv overallTy env tpenv (seqExprOnly, isFromSource, synPat, s
         // optimize 'for i in n .. m do'
         | Expr.App (Expr.Val (vref, _, _), _, [tyarg], [startExpr;finishExpr], _)
              when valRefEq g vref g.range_op_vref && typeEquiv g tyarg g.int_ty ->
-               (g.int32_ty, (fun _ x -> x), id, Choice1Of3 (startExpr, finishExpr))
+               (g.int32_ty, (fun _ x -> x), id, Choice1Of4 (startExpr, finishExpr))
+
+        // optimize 'for i in n .. step .. m do'
+        | Expr.App(Expr.Val(vref, _, _), _, [ tyarg; stepTyarg ], [ startExpr; stepExpr; finishExpr ], _)
+            when valRefEq g vref g.range_step_op_vref && typeEquiv g tyarg g.int_ty && typeEquiv g stepTyarg g.int_ty ->
```

Comment:
I think we may need to guard this via a language version switch or move this to an optimization phase (though that is painful).  It changes the form of quotations for these constructs which is a breaking change.

Quoting these loops is rare but it's also more philosophical: we shouldn't be adding any optimizations which change the quoted form or require new additions to the quotations API.  So optimziations should be in later phases.